content/precise-types.md

```diff
@@ -0,0 +1,523 @@
+  type FooAlias[@precise A] = Foo[A]
+  opaque type FooOpaque[@precise A] = Foo[A]
+object morePreciseErr:
+  trait Foo[T]
+  type FooAlias[@precise A] = Foo[A] // error
+  opaque type FooOpaque[@precise A] = Foo[A] // error
+object lessPreciseErr:
+  trait Foo[@precise T]
+  type FooAlias[A] = Foo[A]
+  opaque type FooOpaque[A] = Foo[A] // error
+~~~
+
+#### Extending Traits & Classes with Precise Params
+A class can extend an existing trait or class without precise arguments and increase the preciseness of its values by applying precise annotations on the relevant type parameters. However, if the extended class or trait already has precise type parameter, the extending class must preserve the same annotation.
+~~~ scala
+object samePreciseOK:
+  trait Foo[@precise T]:
+    val value: T
+  class FooExtend[@precise A](val value: A) extends Foo[A]
+object morePreciseOK:
+  trait Foo[T]:
+    val value: T
+  class FooExtend[@precise A](val value: A) extends Foo[A]
+object lessPreciseErr:
+  trait Foo[@precise T]:
+    val value: T
+  class FooExtend[A](val value: A) extends Foo[A] // error
+~~~ 
+
+#### Typechecking of Polynomial Types with Precise Params
+Signature of polynomial types and values must include precise annotations to match, or else a type-mismatch error is generated.
+~~~ scala
+type Id1 = [@precise T] => T => Box[T]
+val id1Check: Id1 = [T] => (t : T) => Box(t) // error
+
+type Id2 = [T] => T => Box[T]
+val id2Check: Id2 = [@precise T] => (t: T) => Box(t) // error
+~~~
+
+### Compatibility
+Fully backward compatible.
+
+
+### Performance
+
+This feature comes with some performance penalty. Extra logic is added to most typing operations to identify precise types (unless precise mode is already active). Additionally, because methods are selected according to their arguments' imprecise types, arguments may need to be (re)typed precisely as discussed in special condition subsections of the spec.
+
+### Open questions
+
+#### Precise annotation in other places
+
+Currently this proposal supports `@precise` to be applied only on type parameters. Some questions come to mind:
+1. Should we support annotating values and definitions with `@precise`? The semantics of this could be that such values and defs will never have explicit return type annotation and rely on precise typing to set the return type. This is somewhat similar to using `final val` or `final def`, with the advantage that `final` is not available locally and `@precise` can be. See related [feature request](https://github.com/lampepfl/dotty-feature-requests/issues/48). 
+2. Should we support annotating objects and classes with `@precise`? Theoretically we could define that a precise class or object semantically meant that all values and defs in their scope are considered to be precise. This is dependent on supporting `@precise` values and definitions (Question 1). Personally, I think this is bad to enable, but maybe someone can convince me of a good use-case.
```

Comment:
Thanks @soronpo , this looks great!

I think supporting this at the class level may be confusing because of subclassing and how we can depend on third-party base classes. 
If classes are allowed, I think subclasses would need to carry the annotation. If it inferred automatically different rules of typing without being explicit in the annotation, the user might be left wondering what's causing the methods not to type as usual. 

If some code depends on a base class that is not in the local module, the user may not understand why after a release of a library, their @precise typing affects their local scope. They'd have to visit and inspect the dependency sources manually.